src/replit_river/v2/session.py

```diff
@@ -0,0 +1,1239 @@
+            logger.exception(
+                f"Error connecting, retrying with {backoff_time}ms backoff"
+            )
+            await asyncio.sleep(backoff_time / 1000)
+    unbind_connecting_task()
+
+    if last_error is not None:
+        logger.debug("Handshake attempts exhausted, terminating")
+        do_close()
+        raise RiverException(
+            ERROR_HANDSHAKE,
+            f"Failed to create ws after retrying {attempt_count} number of times",
+        ) from last_error
+
+    return None
+
+
+async def _recv_from_ws(
+    block_until_connected: Callable[[], Awaitable[None]],
+    client_id: str,
+    get_state: Callable[[], SessionState],
+    get_ws: Callable[[], ClientConnection | None],
+    transition_no_connection: Callable[[], Awaitable[None]],
+    close_session: Callable[[], Awaitable[None]],
+    assert_incoming_seq_bookkeeping: Callable[
+        [str, int, int], Literal[True] | _IgnoreMessage
+    ],
+    get_stream: Callable[[str], tuple[Channel[Exception | None], Channel[Any]] | None],
+    enqueue_message: SendMessage[None],
+) -> None:
+    """Serve messages from the websocket.
+
+    Process incoming packets from the connected websocket.
+    """
+    our_task = asyncio.current_task()
+    connection_attempts = 0
+    try:
+        while our_task and not our_task.cancelling() and not our_task.cancelled():
+            logger.debug(f"_recv_from_ws loop count={connection_attempts}")
+            connection_attempts += 1
+            ws = None
+            while ((state := get_state()) in ConnectingStates) and (
+                state not in TerminalStates
+            ):
+                logger.debug(
+                    "_handle_messages_from_ws spinning while connecting, %r %r",
+                    ws,
+                    state,
+                )
+                await block_until_connected()
+
+            if state in TerminalStates:
+                logger.debug(
+                    f"Session is {state}, shut down _recv_from_ws",
+                )
+                # session is closing / closed, no need to _recv_from_ws anymore
+                break
+
+            logger.debug("client start handling messages from ws %r", ws)
+
+            error_channel: Channel[Exception | None] | None = None
+
+            # We should not process messages if the websocket is closed.
+            while (ws := get_ws()) and get_state() in ActiveStates:
+                connection_attempts = 0
+                # decode=False: Avoiding an unnecessary round-trip through str
+                # Ideally this should be type-ascripted to : bytes, but there
+                # is no @overrides in `websockets` to hint this.
+                try:
+                    message = await ws.recv(decode=False)
+                except ConnectionClosed:
+                    # This triggers a break in the inner loop so we can get back to
+                    # the outer loop.
+                    await transition_no_connection()
```

Comment:
could this cause a transition from terminal -> no connection?

Reply:
All of the transition functions are guarded:

```python
        async def transition_no_connection() -> None:
            if self._state in TerminalStates:
                return
```